addons/website_sale/static/src/js/website_sale.editor.js

```diff
@@ -47,6 +47,29 @@ WebsiteNewMenu.include({
 });
 });
 
+odoo.define('website_sale.editMenu', function (require) {
+    'use strict';
+
+var WebsiteEditMenu = require('website.editMenu');
+
+WebsiteEditMenu.include({
+    /**
+     * @override
+     */
+    _getContentEditableAreas () {
+        return $(this.savableSelector).not('input, [data-oe-readonly],[data-oe-type="monetary"],[data-oe-many2one-id], [data-oe-field="arch"]:empty').filter((_, el) => {
+            return !$(el).closest('.o_not_editable, .oe_website_sale .products_header').length;
+        }).toArray();
+    },
+    /**
+     * @override
+     */
+    _getReadOnlyAreas () {
+        return $("#wrapwrap").find('.oe_website_sale .products_header, .oe_website_sale .products_header a').toArray();
+    },
+});
+});
```

Comment:
@dmo-odoo Why do we need this whole bunch of javascript to declare that some area are not editable ? :thinking: Should not it be automatic via the branding added on the HTML and/or the o_not_editable class (well I guess I still have to review this PR https://github.com/odoo/odoo/pull/87347#issuecomment-1090297736 if it is finished... is it @Goaman ? Kinda slipped my mind).

If not, why is there not 500 areas declared in this PR ?

Reply:
@qsm-odoo I was also confused by the naming at first but these are three different features, actually:

- `o_not_editable` is recognized by the editor as a node that cannot be changed by our internal code. However, the browser is unaware of this and may still affect those nodes. This is why we force these nodes to be set as `contentEditable=false` in https://github.com/odoo/odoo/pull/87347#issuecomment-1090297736. While we could have solved the bug here by waiting for that particular PR and adding the `o_not_editable` class, this is stable so you probably don't want to add the `o_not_editable` class in the templates in 15.0 anyway.
- `_getContentEditableAreas` is to set `contentEditable=true` dynamically on elements. Unlike the `o_not_editable` class, this is configurable when instantiating the editor so we can have different editable zones in website, mass_mailing and notes.
- `_getReadonlyAreas` is new and is the opposite of `getContentEditableAreas`. It dynamically sets `contentEditable=false` on elements and is configurable when instantiating the editor.

I agree that the naming is subpar. Notably the "get" made you think that this function should indeed return 500+ elements, while in fact it's more like `getElementsToForcefullySetAsReadonly` or something along those lines. However, since `_getContentEditableAreas` already existed, it made sense to have the opposite one follow the same naming conventions, even if they were not great to begin with.

We could, in master, rename these functions and add some `o_not_editable` classes to parts of the templates that we know will never make sense to be editable and therefore would not need to be dynamically provided via `getReadonlyAreas`.

Reply:
@dmo-odoo @sebgeelen 

> this is stable so you probably don't want to add the o_not_editable class in the templates in 15.0 anyway

You do want that actually IMO. **Adding** a class has always been judged stable (regarding potential xpath custo which should use `hasclass`) and that's probably not a critical fix so it would not be a big deal that existing users do not receive it. However, if we decided to add javascript for this specific case as a stable fix (which would be good, probably user report ?)... why is it not indicated with a TODO comment to adapt in master and why was it forward-ported as it is in master?

> _getContentEditableAreas is to set contentEditable=true dynamically on elements. Unlike the o_not_editable class, this is configurable when instantiating the editor so we can have different editable zones in website, mass_mailing and notes.

It cannot make sense to have this in website:

![image](https://user-images.githubusercontent.com/10338094/169789194-814d590e-c46f-4eec-9ba6-c5c8128ab437.png)

And this in website_sale:

![image](https://user-images.githubusercontent.com/10338094/169789321-22622f4d-8a92-423b-9a08-378434275249.png)

This is obviously duplicated, even the missing spaces in the selector are duplicated. And this showcases even more that adding the `o_not_editable` class at the right place it what is needed here (and again I would not mind a JS solution on top of that o_not_editable addition so that the fix is received by everyone in stable... but not with duplicated and uncommented code).

btw: seems strange that `[data-oe-readonly]` is part of the selector which searches editable areas but I don't know about this well :thinking:

> _getReadonlyAreas is new and is the opposite of getContentEditableAreas. It dynamically sets contentEditable=false on elements and is configurable when instantiating the editor.

So basically the same as o_not_editable after the mentioned fix PR, right? Why not but that seems overly complicated for each module to make an include of an editor class to override a method to list all non editable areas of their application. If the o_not_editable system (which is not perfect as requires to add more info to download for all visitors) is to be replaced, a registry of some kind would be prettier IMO... but in this case I feel that the cost of altering the DOM with o_not_editable is very small and that system is way more easy to use IMO, but it is has the disadvantage of not allowing stable fixes for existing users, that's true. I guess that could co-exist but would need comments etc. Also probably needed to make sure that all places relying on o_not_editable now also relies on that new system **or** that all elements which are declared with that new system receive the o_not_editable class on editor start :shrug: 

> I agree that the naming is subpar. Notably the "get" made you think that this function should indeed return 500+ elements, while in fact it's more like getElementsToForcefullySetAsReadonly

That may be part of what I don't understand: I don't think there are one exception where we want to force something, there are many. There is an automatic system (the branding) which allows to find the areas that *can* be edited. This is what seems to be basically returned by `_getContentEditableAreas`, thanks to the `savableSelector`\*\*. Then there are exceptions where this is technically possible to save... but we don't want the user to do so, so we add the o_not_editable class. Same as explained above: adding another system to forcefully set contenteditable="false" on some other elements that those which have the o_not_editable class may be fine, but if needed it should mention that other system via comments :grimacing: 

\*\* Which btw is defined with dead code I think:
![image](https://user-images.githubusercontent.com/10338094/169793550-f1d2dfb3-7842-41e7-a043-169a6e353d0c.png)

`oeStructureSelector` is a subset of `oeFieldSelector` and `oeStructureSelector` is not used anywhere else that in that chunk of code. Maybe something to review in master? cc @Zinston

Reply:
I discussed with @sebgeelen and indeed the reason he chose to go with redefining these functions was that, in the absence of https://github.com/odoo/odoo/pull/87347, adding the `o_not_editable` class did _not_ fix the bug during the exploration phase when he tried to find a proper fix.

It did not bother me for the reasons I explained in my previous comment, that is that I find the actual mechanics of `_getContentEditableAreas` and `_getReadOnlyAreas` useful. Since I genuinely did not know _adding_ a class in a template in stable fell under a different rule than other template changes, it felt neat to use those mechanics rather than change the template.

However, I do agree that the duplicated override is not ideal. It could not be inferred from the PR changes themselves so thanks for bringing it up.

In light of this, I suggest to wait for https://github.com/odoo/odoo/pull/87347 after which we will remove the website_sale override introduced by this PR and add the `o_not_editable` class in 15.0.

Reply:
@dmo-odoo @Goaman 

> In light of this, I suggest to wait for https://github.com/odoo/odoo/pull/87347 after which we will remove the website_sale override introduced by this PR and add the o_not_editable class in 15.0.

I guess, although it would need to be removed in master only now that it was merged in stable, so adding o_not_editable in 15.0 has no real advantage anymore and could also only be done in master.

Is #87347 ready or not? Should I have a look? @Goaman ? :+1: 
I think #87437 should be adapted to handle what I suggested with:

> I guess that could co-exist but would need comments etc. Also probably needed to make sure that all places relying on o_not_editable now also relies on that new system or that all elements which are declared with that new system receive the o_not_editable class on editor start :shrug: